File: 987-vertical-order-traversal-of-a-binary-tree/987-vertical-order-traversal-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    static Map<Integer, List<Integer>> DEFAULT = new HashMap<>();
    Map<Integer, Map<Integer, List<Integer>>> store;
    int minX, minY, maxX, maxY;
    
    public List<List<Integer>> verticalTraversal(TreeNode root) {
        store = new HashMap<>();
        minX = 0; maxX = 0; minY = 0; maxY = 0;
        res(root, 0, 0);
        List<List<Integer>> ret = new ArrayList<>();
        for (int j = minY; j <= maxY; j++) {
            List<Integer> curr = new ArrayList<>();
            for (int i = minX; i <= maxX; i++) {
                if (store.getOrDefault(i, DEFAULT).containsKey(j)) {
                    Collections.sort(store.get(i).get(j));
                    curr.addAll(store.get(i).get(j));
                }
            }
            if (curr.size() != 0) {
                ret.add(curr);
            }
        }
        return ret;
    }
    
    private void res(TreeNode node, int x, int y) {
        if (node != null) {
            if (!store.containsKey(x)) {
                store.put(x, new HashMap<>());
            }
            if (!store.get(x).containsKey(y)) {
                store.get(x).put(y, new ArrayList());
            }
            store.get(x).get(y).add(node.val);
            minX = Math.min(minX, x);
            minY = Math.min(minY, y);
            maxX = Math.max(maxX, x);
            maxY = Math.max(maxY, y);
            res(node.left, x + 1, y - 1);
            res(node.right, x + 1, y + 1);
        }
    }
}

File: 987-vertical-order-traversal-of-a-binary-tree/987-vertical-order-traversal-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalTraversal(self, root: TreeNode) -> List[List[int]]:
        ret = dict()
        mn = math.inf
        def res(node = root, ver = 0, hei = 0):
            nonlocal mn
            if node is not None:
                if ver in ret and hei in ret[ver]:
                    l = ret[ver][hei]
                    l.append(node.val)
                    i = len(l)-1
                    while i != 0 and l[i] < l[i-1]:
                        l[i], l[i-1] = l[i-1], l[i]
                        i -= 1
                elif ver in ret:
                    ret[ver][hei] = [node.val]
                else:
                    ret[ver] = {hei:[node.val]}
                    mn = min(ver, mn)
                res(node.left, ver-1, hei+1)
                res(node.right, ver+1, hei+1)
        res()
        fin = []
        keys = sorted(list(ret.keys()))
        for i in keys:
            l = []
            innerkeys = sorted(list(ret[i].keys()))
            for j in innerkeys:
                l += ret[i][j]
            fin.append(l)
        return fin
            
            
        